### PR TITLE
SwissDict

### PR DESCRIPTION
**Pull request by eulerkochy**
### SwissDict
Part of GSoC'20 work.
Initial code and tests.
Tests run and works.
As of now, the benchmarks aren't favorable.
I'll post the benchmarks once I finish the work on #517 

**Tasks remaining**

- [x] Refactor code, and improve benchmarks.
- [x] Add docstrings and deploy doc
- [x] Add file to README

### CODE REVIEW

File: src/swiss_dict.jl [hunk 1]
```diff
@@ -0,0 +1,559 @@
+import Base: setindex!, sizehint!, empty!, isempty, length, copy, empty,
+             getindex, getkey, haskey, iterate, @propagate_inbounds, ValueIterator,
+             pop!, delete!, get, get!, isbitstype, in, hashindex, isbitsunion,
+             isiterable, dict_with_eltype, KeySet, Callable, _tablesz, filter!
```
Comment (oxinabox) [inline src/swiss_dict.jl:4]: Do we always do this per file?
I thought we just did this at the top level file

File: src/swiss_dict.jl [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+    end
+    return h
+end
+# Note the constructors of WeakKeyDict mirror these here, keep in sync.
+SwissDict() = SwissDict{Any,Any}()
+SwissDict(kv::Tuple{}) = SwissDict()
+copy(d::SwissDict) = SwissDict(d)
+empty(d::SwissDict, ::Type{K}, ::Type{V}) where {K, V} = SwissDict{K, V}()
+
+const AnyDict = SwissDict{Any,Any}
```
Comment (oxinabox) [inline src/swiss_dict.jl:50]: this is never used and seems unuseful and confusing

File: src/swiss_dict.jl [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+            merge!(new, dest)
+            new[k] = v
+            return grow_to!(new, itr, st)
+        end
+        y = iterate(itr, st)
+    end
+    return dest
+end
+
+empty(a::AbstractDict, ::Type{K}, ::Type{V}) where {K, V} = SwissDict{K, V}()
```
Comment (oxinabox) [inline src/swiss_dict.jl:95]: Why are we chanign the default definition of `empty` for all `AbstractDicts`?
That doesn't seem right

File: src/swiss_dict.jl [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+
+#hashindex(key, sz) = (((hash(key)%Int) & (sz-1)) + 1)::Int
+
+##Simd utilities
```
Comment (oxinabox) [inline src/swiss_dict.jl:99]:
```suggestion
##SIMD utilities
```

Also fun, i haven't touched these `llvmcall` for SIMD in ages.

File: src/swiss_dict.jl [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+    ret void
+    """), Nothing, Tuple{UInt}, p)
+    else
+        Core.Intrinsics.llvmcall(("declare void @llvm.prefetch(i8*, i32, i32, i32)", """
+    %ptr = inttoptr i32 %0 to i8*
+    call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
+    ret void
+    """), Nothing, Tuple{UInt}, p)
+    end
+end
```
Comment (oxinabox) [inline src/swiss_dict.jl:132]: Can we get rid of the `if` by interpolating the type string in?

```suggestion
@inline function _prefetchr(p::Ptr)
    p = reinterpret(UInt, p)
    int_t = Int === Int64 ? "i64" : "int_t"
    Core.Intrinsics.llvmcall(
        ("declare void @llvm.prefetch(i8*, i32, i32, i32)",
        """
        %ptr = inttoptr $int_t %0 to i8*
        call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
        ret void
        """),
        Nothing, Tuple{UInt}, p
    )
end
```

File: src/swiss_dict.jl [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+
+@inline function _prefetchw(p::Ptr)
+    p = reinterpret(UInt, p)
+    if UInt === UInt64
```
Comment (oxinabox) [inline src/swiss_dict.jl:136]: similar to above, can we interpolate the type string in?

File: src/swiss_dict.jl [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+        hi = ((u>>8) | (u<<56)) % Int
+    else
+        hi = ((u>>8) | (u<<24)) % Int
+    end
+    tag = tag > 1 ? tag : tag+0x02
+    return (hi, tag)
+end
+
+@propagate_inbounds function _slotget(slots::Vector{_u8x16}, i::Int)
+    #@boundscheck 0 < i <= length(slots)*16 || throw(BoundsError(slots, 1 + (i-1)>>4)
```
Comment (oxinabox) [inline src/swiss_dict.jl:165]: why is this commented  out?

File: test/runtests.jl [hunk 8]
```diff
@@ -27,10 +27,11 @@ tests = ["int_set",
          "multi_dict",
          "circular_buffer",
          "sorting",
-         "priority_queue",
-         "fenwick",
+         "priority_queue", 
+         "fenwick",
```
Comment (oxinabox) [inline test/runtests.jl:31]: I think maybe you did not intend to add this white-space?

**Comment by rfourquet** [comment]
Could you post a link to a reference on what are "Swiss dictionaries" ? It's difficult to find relevant results on search engines with these terms, also with "swiss map" ;-)

**Comment by eulerkochy** [comment]
@rfourquet `SwissDict` as such doesn't exist. It is based on Abseil's flat_hash_map, which in turn uses [Swiss Tables](https://abseil.io/blog/20180927-swisstables), develeped by Google engineers. Here's the [cppcon](https://www.youtube.com/watch?v=ncHmEUmJZf4) talk by Matt Kulukundis which inspired my implementation.

**Comment by rfourquet** [comment]
@eulerkochy Thanks!

File: src/swiss_dict.jl (hunk 1 above)
Comment (eulerkochy) [reply src/swiss_dict.jl:4]: it's there in `robin_dict.jl`. Nice suggestion, let me get to it and try to get it out from there as well. I'll remove this from here as well.

File: src/swiss_dict.jl [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+    call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
+    ret void
+    """), Nothing, Tuple{UInt}, p)
+    else
+        Core.Intrinsics.llvmcall(("declare void @llvm.prefetch(i8*, i32, i32, i32)", """
+    %ptr = inttoptr i32 %0 to i8*
+    call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
+    ret void
+    """), Nothing, Tuple{UInt}, p)
+    end
```
Comment (vtjnash) [inline src/swiss_dict.jl:98]:
```suggestion
        ccall("llvm.prefetch", llvmcall, Cvoid, (Ref{Int8}, Int32, Int32, Int32), Ptr{Int8}(p), 0, 3, 1)
    end
```

**Comment by eulerkochy** [comment]
| K=Int64, 1e4 elements  | Dict                      | SwissDict                 | RobinDict                 |
|------------------------|---------------------------|---------------------------|---------------------------|
| make                   | 364.609375 kB /283.927 μs | 364.453125 kB /348.340 μs | 428.421875 kB /413.592 μs |
| pop!                   | 30.000 ns                 | 44.000 ns                 | 56.000 ns                 |
| empty!                 | 339.000 ns                | 658.000 ns                | 1.437 μs                  |
| find-success           | 40.000 ns                 | 43.000 ns                 | 40.000 ns                 |
| find-failure           | 40.000 ns                 | 36.000 ns                 | 63.000 ns                 |

File: src/swiss_dict.jl [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+function _setindex!(h::SwissDict, v, key, index, tag)
+    @inbounds h.keys[index] = key
+    @inbounds h.vals[index] = v
+    h.count += 1
+    h.age += 1
+    so =  _slotget(h.slots, index)
+    h.nbfull += (iszero(index & 0x0f) & (so==0x00))
+    _slotset!(h.slots, tag, index)
+    if index < h.idxfloor
+        h.idxfloor = index
```
Comment (oxinabox) [inline src/swiss_dict.jl:218]: do we need to fix the tests to make sure to cover this?

**Review by oxinabox — APPROVED** [review]

File: src/swiss_dict.jl (hunk 10 above)
Comment (eulerkochy) [reply src/swiss_dict.jl:218]: I can't seem to find a test to cover this. It's quite hard to find a test case specifically for this. But, this serves the purpose of `h.idxfloor`, so it isn't wrong.

**Pull request merged**